### PR TITLE
chore: update estree-walker to v3

### LINKED ISSUE #3451: Update dependencies that require ESM
Right now we have several dependencies that we can't update, because they [rely on ES Modules rather than CommonJS](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c):

- [x] `parse5`
- [ ] `estree-walker`
- [x] `webdriverio`

(For `webdriverio` it's less severe, since it's only a `devDep`.)

I see two options for solving this:

1. #3017
2. Bundle these dependencies into our `dist` files (e.g. with Rollup)

#<span></span>1 would be the simplest solution, but it's a breaking change. And I'm not sure how realistic it is, given Jest's [lack of mainstream support for ESM](https://jestjs.io/docs/ecmascript-modules). (E.g. would it even be possible to import `@lwc/compiler` into a `jest.config.js` file, if it's `"type": "module"`?)

#<span></span>2 is kind of gross, but may be the most pragmatic solution. We can continue to ship both CJS and ESM as we do today, without breaking downstream consumers of our npm package.

### PR DESCRIPTION
**Pull request by nolanlawson**
## Details

Fixes #3451

Updates `estree-walker` to v3, which is a breaking change because it now requires ESM.

Similar to how we solved `parse5`, we solve this by simply bundling `estree-walker` into our `dist `files. This avoids shipping a breaking change to our downstream consumers.

## Does this pull request introduce a breaking change?

-   😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

-   🤞 No, it does not introduce an observable change.

### CODE REVIEW

File: packages/@lwc/template-compiler/package.json
Comment (nolanlawson) [inline packages/@lwc/template-compiler/package.json:56]: Moving it to `devDeps` allows it to be inlined

File: packages/@lwc/template-compiler/src/codegen/codegen.ts
Comment (nolanlawson) [inline packages/@lwc/template-compiler/src/codegen/codegen.ts:556]: This comment is copied from:

https://github.com/salesforce/lwc/blob/8a0f98fb42aa0db38d0148e402a7a743e3ad7e56/packages/%40lwc/template-compiler/src/shared/types.ts#L65-L69

**Comment by wjhsf** [comment]
Does their license allow us to do this? Do we need to file a new 3PP approval since we're distributing it ourselves, now?

**Review by wjhsf — APPROVED** [review]

**Comment by nolanlawson** [comment]
> Does their license allow us to do this? Do we need to file a new 3PP approval since we're distributing it ourselves, now?

Technically we should have figured this out when we first started bundling. But anyway I opened an issue to track this: https://github.com/salesforce/lwc/issues/4133

**Comment by nolanlawson** [comment]
OK I guess this should wait until #4134 is merged.

**Pull request merged**
